File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter14/FingerPaint/FingerPaint/XYSlider.cpp
//
// XYSlider.xaml.cpp
// Implementation of the XYSlider class
//

#include "pch.h"
#include "XYSlider.h"

using namespace FingerPaint;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;

DependencyProperty^ XYSlider::valueProperty = 
    DependencyProperty::Register("Point", 
                                 TypeName(Point::typeid), 
                                 TypeName(XYSlider::typeid), 
                                 ref new PropertyMetadata(Point(0.5f, 0.5f), 
                                 ref new PropertyChangedCallback(&XYSlider::OnDependencyPropertyValueChanged)));

XYSlider::XYSlider()
{
    this->DefaultStyleKey = "XYSliderDemo.XYSlider";
    pointerDictionary = ref new Map<unsigned int, Point>();
}

void XYSlider::OnApplyTemplate()
{
    // Detach event handlers
    if (contentPresenter != nullptr)
    {
        contentPresenter->PointerPressed -= contentPresenterPointerPressedEventToken;
        contentPresenter->PointerMoved -= contentPresenterPointerMovedEventToken;
        contentPresenter->PointerReleased -= contentPresenterPointerReleasedEventToken;
        contentPresenter->PointerCaptureLost -= contentPresenterPointerCaptureLostEventToken;

        contentPresenter->SizeChanged -= contentPresenterSizeChangedEventToken;
    }

    // Get new parts
    crossHairPart = dynamic_cast<FrameworkElement^>(GetTemplateChild("CrossHairPart"));
    contentPresenter = dynamic_cast<ContentPresenter^>(GetTemplateChild("ContentPresenterPart"));

    // Attach event handlers
    if (contentPresenter != nullptr)
    {
        contentPresenter->ManipulationMode = ManipulationModes::TranslateX | ManipulationModes::TranslateY;

        contentPresenterPointerPressedEventToken = contentPresenter->PointerPressed += ref new PointerEventHandler(this, &XYSlider::OnContentPresenterPointerPressed);
        contentPresenterPointerMovedEventToken = contentPresenter->PointerMoved += ref new PointerEventHandler(this, &XYSlider::OnContentPresenterPointerMoved);
        contentPresenterPointerReleasedEventToken = contentPresenter->PointerReleased += ref new PointerEventHandler(this, &XYSlider::OnContentPresenterPointerReleased);
        contentPresenterPointerCaptureLostEventToken = contentPresenter->PointerCaptureLost += ref new PointerEventHandler(this, &XYSlider::OnContentPresenterPointerReleased);

        contentPresenterSizeChangedEventToken = contentPresenter->SizeChanged += ref new SizeChangedEventHandler(this, &XYSlider::OnContentPresenterSizeChanged);
    }

    // Make cross-hair transparent to touch
    if (crossHairPart != nullptr)
    {
        crossHairPart->IsHitTestVisible = false;
    }
}

void XYSlider::OnContentPresenterPointerPressed(Object^ sender, PointerRoutedEventArgs^ args)
{
    unsigned int id = args->Pointer->PointerId;
    Point point = args->GetCurrentPoint(contentPresenter)->Position;
    pointerDictionary->Insert(id, point);
    contentPresenter->CapturePointer(args->Pointer);

    RecalculateValue();
}

void XYSlider::OnContentPresenterPointerMoved(Object^ sender, PointerRoutedEventArgs^ args)
{
    unsigned int id = args->Pointer->PointerId;

    if (pointerDictionary->HasKey(id))
    {
        Point point = args->GetCurrentPoint(contentPresenter)->Position;
        pointerDictionary->Insert(id, point);
        RecalculateValue();
    }
}

void XYSlider::OnContentPresenterPointerReleased(Object^ sender, PointerRoutedEventArgs^ args)
{
    unsigned int id = args->Pointer->PointerId;

    if (pointerDictionary->HasKey(id))
    {
        pointerDictionary->Remove(id);
        RecalculateValue();
    }
}

void XYSlider::OnContentPresenterSizeChanged(Object^ sender, SizeChangedEventArgs^ args)
{
    SetCrossHair();
}

void XYSlider::RecalculateValue()
{
    if (pointerDictionary->Size > 0)
    {
        Point accumPoint(0, 0);

        // Average all teh curent touch points
        for each(IKeyValuePair<unsigned int, Point>^ pair in pointerDictionary)
        {
            accumPoint.X += pair->Value.X;
            accumPoint.Y += pair->Value.Y;
        }

        accumPoint.X /= pointerDictionary->Size;
        accumPoint.Y /= pointerDictionary->Size;

        RecalculateValue(accumPoint);
    }
}

void XYSlider::RecalculateValue(Point absolutePoint)
{
    double x = max(0.0, min(1.0, absolutePoint.X / contentPresenter->ActualWidth));
    double y = max(0.0, min(1.0, absolutePoint.Y / contentPresenter->ActualHeight));
    this->Value = Point((float)x, (float)y);
}

void XYSlider::SetCrossHair()
{
    if (contentPresenter != nullptr && crossHairPart != nullptr)
    {
        Canvas::SetLeft(crossHairPart, this->Value.X * contentPresenter->ActualWidth);
        Canvas::SetTop(crossHairPart, this->Value.Y * contentPresenter->ActualHeight);
    }
}

void XYSlider::OnDependencyPropertyValueChanged(Windows::UI::Xaml::DependencyObject^ obj, Windows::UI::Xaml::DependencyPropertyChangedEventArgs^ args)
{
    dynamic_cast<XYSlider^>(obj)->SetCrossHair();
    dynamic_cast<XYSlider^>(obj)->OnValueChanged((Point)args->NewValue);
}

void XYSlider::OnValueChanged(Point value)
{
    ValueChanged(this, value);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter17/FingerPaint/FingerPaint/XYSlider.cs
using System;
using System.Collections.Generic;
using Windows.Foundation;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Input;

namespace FingerPaint
{
    public class XYSlider : ContentControl
    {
        ContentPresenter contentPresenter;
        FrameworkElement crossHairPart;
        Dictionary<uint, Point> pointerDictionary = new Dictionary<uint, Point>();

        static readonly DependencyProperty valueProperty = 
                DependencyProperty.Register("Value",
                        typeof(Point), typeof(XYSlider), 
                        new PropertyMetadata(new Point(), OnValueChanged));

        public event EventHandler<Point> ValueChanged;

        public XYSlider()
        {
            this.DefaultStyleKey = typeof(XYSlider);
        }

        public static DependencyProperty ValueProperty
        {
            get { return valueProperty; }
        }

        public Point Value
        {
            set { SetValue(ValueProperty, value); }
            get { return (Point)GetValue(ValueProperty); }
        }

        protected override void OnApplyTemplate()
        {
            // Detach event handlers
            if (contentPresenter != null)
            {
                contentPresenter.PointerPressed -= OnContentPresenterPointerPressed;
                contentPresenter.PointerMoved -= OnContentPresenterPointerMoved;
                contentPresenter.PointerReleased -= OnContentPresenterPointerReleased;
                contentPresenter.PointerCaptureLost -= OnContentPresenterPointerReleased;
                contentPresenter.SizeChanged -= OnContentPresenterSizeChanged;
            }

            // Get new parts
            crossHairPart = GetTemplateChild("CrossHairPart") as FrameworkElement;
            contentPresenter = GetTemplateChild("ContentPresenterPart") as ContentPresenter;

            // Attach event handlers
            if (contentPresenter != null)
            {
                contentPresenter.PointerPressed += OnContentPresenterPointerPressed;
                contentPresenter.PointerMoved += OnContentPresenterPointerMoved;
                contentPresenter.PointerReleased += OnContentPresenterPointerReleased;
                contentPresenter.PointerCaptureLost += OnContentPresenterPointerReleased;
                contentPresenter.SizeChanged += OnContentPresenterSizeChanged;
            }

            // Make cross-hair transparent to touch
            if (crossHairPart != null)
            {
                crossHairPart.IsHitTestVisible = false;
            }

            base.OnApplyTemplate();
        }

        void OnContentPresenterPointerPressed(object sender, PointerRoutedEventArgs args)
        {
            uint id = args.Pointer.PointerId;
            Point point = args.GetCurrentPoint(contentPresenter).Position;
            pointerDictionary.Add(id, point);
            contentPresenter.CapturePointer(args.Pointer);

            RecalculateValue();
            args.Handled = true;
        }

        void OnContentPresenterPointerMoved(object sender, PointerRoutedEventArgs args)
        {
            uint id = args.Pointer.PointerId;
            Point point = args.GetCurrentPoint(contentPresenter).Position;

            if (pointerDictionary.ContainsKey(id))
            {
                pointerDictionary[id] = point;
                RecalculateValue();
                args.Handled = true;
            }
        }

        void OnContentPresenterPointerReleased(object sender, PointerRoutedEventArgs args)
        {
            uint id = args.Pointer.PointerId;

            if (pointerDictionary.ContainsKey(id))
            {
                pointerDictionary.Remove(id);
                RecalculateValue();
                args.Handled = true;
            }
        }

        void OnContentPresenterSizeChanged(object sender, SizeChangedEventArgs args)
        {
            SetCrossHair();
        }

        void RecalculateValue()
        {
            if (pointerDictionary.Values.Count > 0)
            {
                Point accumPoint = new Point();

                // Average all the current touch points
                foreach (Point point in pointerDictionary.Values)
                {
                    accumPoint.X += point.X;
                    accumPoint.Y += point.Y;
                }
                accumPoint.X /= pointerDictionary.Values.Count;
                accumPoint.Y /= pointerDictionary.Values.Count;

                RecalculateValue(accumPoint);
            }
        }

        void RecalculateValue(Point absolutePoint)
        {
            double x = Math.Max(0, Math.Min(1, absolutePoint.X / contentPresenter.ActualWidth));
            double y = Math.Max(0, Math.Min(1, absolutePoint.Y / contentPresenter.ActualHeight));
            this.Value = new Point(x, y);
        }

        void SetCrossHair()
        {
            if (contentPresenter != null && crossHairPart != null)
            {
                Canvas.SetLeft(crossHairPart, this.Value.X * contentPresenter.ActualWidth);
                Canvas.SetTop(crossHairPart, this.Value.Y * contentPresenter.ActualHeight);
            }
        }

        static void OnValueChanged(DependencyObject obj, DependencyPropertyChangedEventArgs args)
        {
            (obj as XYSlider).SetCrossHair();
            (obj as XYSlider).OnValueChanged((Point)args.NewValue);
        }

        protected void OnValueChanged(Point value)
        {
            if (ValueChanged != null)
                ValueChanged(this, value);
        }
    }
}
